File: Coocoo3DGraphics/Texture2D.cpp
#include "pch.h"
#include "Texture2D.h"
#include "DirectXHelper.h"
using namespace Coocoo3DGraphics;

void Texture2D::ReloadAsDepthStencil(int width, int height, Format format)
{
	this->width = width;
	this->height = height;
	if ((DXGI_FORMAT)format == DXGI_FORMAT_D24_UNORM_S8_UINT)
		this->m_format = DXGI_FORMAT_R24_UNORM_X8_TYPELESS;
	else if ((DXGI_FORMAT)format == DXGI_FORMAT_D32_FLOAT)
		this->m_format = DXGI_FORMAT_R32_FLOAT;
	this->m_dsvFormat = (DXGI_FORMAT)format;
	this->m_rtvFormat = DXGI_FORMAT_UNKNOWN;
	this->mipLevels = 1;
}

void Texture2D::ReloadAsRenderTarget(int width, int height, Format format)
{
	this->width = width;
	this->height = height;
	this->m_format = (DXGI_FORMAT)format;
	this->m_dsvFormat = DXGI_FORMAT_UNKNOWN;
	this->m_rtvFormat = (DXGI_FORMAT)format;
	this->m_uavFormat = DXGI_FORMAT_UNKNOWN;
	this->mipLevels = 1;
}

void Texture2D::ReloadAsRTVUAV(int width, int height, Format format)
{
	this->width = width;
	this->height = height;
	this->m_format = (DXGI_FORMAT)format;
	this->m_dsvFormat = DXGI_FORMAT_UNKNOWN;
	this->m_rtvFormat = (DXGI_FORMAT)format;
	this->m_uavFormat = (DXGI_FORMAT)format;
	this->mipLevels = 1;
}

void Texture2D::Reload(Texture2D^ texture)
{
	this->width = texture->width;
	this->height = texture->height;
	this->resource = texture->resource;
	this->mipLevels = texture->mipLevels;
	this->Status = texture->Status;
}

void Texture2D::Unload()
{
	width = 0;
	height = 0;
	resource.Reset();
	Status = GraphicsObjectStatus::unload;
}

Format Texture2D::GetFormat()
{
	if (m_dsvFormat != DXGI_FORMAT_UNKNOWN)
		return (Format)m_dsvFormat;
	return (Format)m_format;
}

void Texture2D::StateTransition(ID3D12GraphicsCommandList* commandList, D3D12_RESOURCE_STATES state)
{
	if (prevResourceState != state)
		commandList->ResourceBarrier(1, &CD3DX12_RESOURCE_BARRIER::Transition(resource.Get(), prevResourceState, state));
	prevResourceState = state;
}

File: Coocoo3DGraphics1/Texture2D.cs
using System;
using System.Collections.Generic;
using System.Text;
using Vortice.Direct3D12;
using Vortice.DXGI;
using static Coocoo3DGraphics.DXHelper;

namespace Coocoo3DGraphics
{
    public class Texture2D : IDisposable
    {
        public ID3D12Resource resource;
        public string Name;
        public ResourceStates resourceStates;
        public int width;
        public int height;
        public int mipLevels;
        public Format format;
        public Format rtvFormat;
        public Format dsvFormat;
        public Format uavFormat;
        public GraphicsObjectStatus Status;

        public void ReloadAsDepthStencil(int width, int height, Format format)
        {
            this.width = width;
            this.height = height;
            if (format == Format.D24_UNorm_S8_UInt)
                this.format = Format.R24_UNorm_X8_Typeless;
            else if (format == Format.D32_Float)
                this.format = Format.R32_Float;
            this.dsvFormat = format;
            this.rtvFormat = Format.Unknown;
            this.mipLevels = 1;
        }

        public void ReloadAsRenderTarget(int width, int height, Format format)
        {
            this.width = width;
            this.height = height;
            this.format = format;
            this.dsvFormat = Format.Unknown;
            this.rtvFormat = format;
            this.uavFormat = Format.Unknown;
            this.mipLevels = 1;
        }


        public void ReloadAsRTVUAV(int width, int height, Format format) => ReloadAsRTVUAV(width, height, 1, format);
        public void ReloadAsRTVUAV(int width, int height, int mipLevels, Format format)
        {
            this.width = width;
            this.height = height;
            this.format = format;
            this.dsvFormat = Format.Unknown;
            this.rtvFormat = format;
            this.uavFormat = format;
            this.mipLevels = mipLevels;
        }

        public void Reload(Texture2D texture)
        {
            this.width = texture.width;
            this.height = texture.height;
            this.resource = texture.resource;
            this.mipLevels = texture.mipLevels;
            this.Status = texture.Status;
        }

        public void StateChange(ID3D12GraphicsCommandList commandList, ResourceStates states)
        {
            if (states != resourceStates)
            {
                commandList.ResourceBarrierTransition(resource, resourceStates, states);
                resourceStates = states;
            }
            else if (states == ResourceStates.UnorderedAccess)
            {
                commandList.ResourceBarrierUnorderedAccessView(resource);
            }
        }

        public Format GetFormat()
        {
            if (dsvFormat != Format.Unknown)
                return dsvFormat;
            return format;
        }

        public void Dispose()
        {
            resource?.Release();
            resource = null;
        }
    }
}
